hir-macros/src/lib.rs

```diff
@@ -25,6 +26,71 @@ pub fn derive_spanned(input: proc_macro::TokenStream) -> proc_macro::TokenStream
     }
 }
 
+/// #[operation(
+///     dialect = HirDialect,
+///     traits(Terminator),
+///     implements(BranchOpInterface),
+/// )]
+/// pub struct Switch {
+///     #[operand]
+///     selector: UInt32,
+///     #[successors(keyed)]
+///     cases: SwitchArm,
+///     #[successor]
+///     fallback: Successor,
+/// }
+///
+/// pub struct Call {
+///     #[attr]
+///     callee: Symbol,
+///     #[operands]
+///     arguments: Vec<AnyType>,
+///     #[results]
+///     results: Vec<AnyType>,
+/// }
+///
+/// #[operation]
+/// pub struct If {
+///     #[operand]
+///     condition: Bool,
+///     #[region]
+///     then_region: RegionRef,
+///     #[region]
+///     else_region: RegionRef,
+/// }
+#[proc_macro_attribute]
+pub fn operation(
```

Comment:
I really dig the `operation` proc macro attribute. I expanded it in a few ops. So much pretty complex boilerplate is generated. Awesome!

Reply:
I don't know if you say the `derive!` macro _before_ I implemented the `#[operation]` attribute macro, but...it was madness 😅.

Now it is much cleaner (and more maintainable!), but it does a lot under the covers, so the main tradeoff is lack of visibility into all of that. That said, I tried to keep the macro focused on generating boilerplate for things that feel pretty intuitive, e.g. deriving traits, op construction and verification, correspondence between fields of the struct definition and the methods that get generated. So hopefully we never need to actually look at the generated code except in rare circumstances.